### DISCUSSION TITLE
Providing Initial Session from localstorage/cookie

### DISCUSSION THREAD

**Discussion by thomasmillercf** [nextauthjs/next-auth#7394] [Help]
Hi 

We are using nextjs for multiple applications under the same domain. 

When we navigate between the pages we want to avoid calling /api/auth/session 

We have tried to cache the session in a cookie and provide that initial session to the SessionProiver however, the server side and the client side render do not match, and we get this error.
```tsx
export const AuthSessionProvider = (props: SessionProviderProps) => {
  const initSession = getCookie(SESSION_COOKIE_NAME) ;
  const session = initSession ? JSON.parse(initSession) : undefined;
  return (
    <SessionProvider {...props} session={session}>
      {props.children}
    </SessionProvider>
  );
};

```

![image](https://user-images.githubusercontent.com/112876429/235173963-bca90da0-2829-45e9-a347-4aa262e48f18.png)

We have also tried rendering the providing with a temporary state such as null or undefined however, once the cookie has been loaded, the session is not updated. 

It would be good if there was a way to put the provider in a loading state while we fetch the initial session.

We have one workaround which involves just not rendering the SessionProvider until we have checked the cache cookie.

```tsx
export const AuthSessionProvider = (props: SessionProviderProps) => {
  const [cachedSession, setCachedSession] = useState<CachedSessionContext>({
    status: 'loading',
  });
  useEffect(() => {
    const retrieveSessionCookie = () => {
      const sessionCookie = getCookie(SESSION_COOKIE_NAME);
      if (sessionCookie) {
        const session = JSON.parse(decodeURIComponent(sessionCookie));
        setCachedSession({
          data: session,
          status: 'authenticated',
        });
      } else {
        setCachedSession({
          status: 'missing',
        });
      }
    };

    retrieveSessionCookie();
  }, []);
  if (cachedSession.status === 'loading') {
    return null;
  }

  return (
    <SessionProvider {...props} session={cachedSession.data}>
      {props.children}
    </SessionProvider>
  );
};
```

**Comment by Thinkscape** [comment]
An alternative using the app router:

1. Fetch the session in your RSC, could be root layout, i.e.

```tsx
// written from memory
export default async function RootLayout({children} : { children: ReactNode }) {
  const session = await getServerSession(); // gets session from db or cookie

  return 
    <MySessionProvider session={session}>
      {children}
    </MySessionProvider>;
}
```

2. Have your MySessionProvider be a client comp. and wrap the next-auth one, something like

```tsx
"use client"
import { SessionProvider } from 'next-auth';

export default function MySessionProvider({ session: Session, children: ReactNode }) {
  return 
    <SessionProvider session={session}>
      {children}
    </Session>
  ;
}
```
3. Use as usual in client components, i.e. 

```tsx
"use client"

export default function Dashboard(){
  const session = useSession();

  return <p>Welcome to dashboard {session.user?.name}</p>
}

```

### The pre-populated SessionProvider bug #7399

There's a bug/feature in `useSession()` and/or SessionProvider that will trigger a session refresh on mount. I've spent some time debugging but couldn't nail it down - something wonky with internal state management (none of the props helped). That means it will be pre-populated, but still send one call to `/api/auth/session`.

The workaround to this bug is to drop the SessionProvider altogether, and use `React.context()`.

Something like

``` tsx
import { type Session } from "next-auth";
import React from "react";

export const SessionContext = React.createContext?.<Session>({});

export default function MySessionProvider({
  children,
  session,
}: {
  children: React.ReactNode;
  session: Session;
}) {
  return (
    <>
      <SessionContext.Provider value={session}>
        {children}
      </SessionContext.Provider>
    </>
  );
```

This way you can populate it once from RSC (i.e. RootLayout) and you're set up. You consume it with `const session = useContext(SessionContext);`. Keep in mind that that example is missing auto-refresh functionality, so if session expires, it'll behave like it's still there. You can either implement it in the `MySessionProvider` or (IMHO better yet) in the your API client (when API calls start failing, refresh session or redirect to login).

Hoping it helps.
